qupath-gui-fx/src/main/java/qupath/lib/gui/scripting/ScriptEditorDragDropListener.java

```diff
@@ -0,0 +1,80 @@
+/*-
+ * #%L
+ * This file is part of QuPath.
+ * %%
+ * Copyright (C) 2014 - 2016 The Queen's University of Belfast, Northern Ireland
+ * Contact: IP Management (ipmanagement@qub.ac.uk)
+ * Copyright (C) 2018 - 2022 QuPath developers, The University of Edinburgh
+ * %%
+ * QuPath is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ * 
+ * QuPath is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License 
+ * along with QuPath.  If not, see <https://www.gnu.org/licenses/>.
+ * #L%
+ */
+
+package qupath.lib.gui.scripting;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javafx.event.EventHandler;
+import javafx.scene.input.DragEvent;
+import javafx.scene.input.Dragboard;
+import qupath.lib.gui.QuPathGUI;
+import qupath.lib.gui.scripting.languages.ScriptLanguageProvider;
+
+/**
+ * Drag and drop support for QuPath's script editor, which can support a range of different files (Plain text, JSON, Groovy,..)
+ * @author Melvin Gelbard
+ */
+public class ScriptEditorDragDropListener implements EventHandler<DragEvent> {
+
+	private final QuPathGUI qupath;
+
+	/**
+	 * Constructor.
+	 * @param qupath the current QuPath instance
+	 */
+	public ScriptEditorDragDropListener(final QuPathGUI qupath) {
+		this.qupath = qupath;
+	}
+
+	@Override
+	public void handle(DragEvent event) {
+		Dragboard dragboard = event.getDragboard();
+        
+		if (dragboard.hasFiles()) {
+			var list = dragboard.getFiles();
+			if (list == null)
+				return;
+			
+			List<File> jars = list.stream().filter(f -> f.getName().toLowerCase().endsWith(".jar")).collect(Collectors.toList());
+			if (!jars.isEmpty() && qupath.canInstallExtensions())
+				qupath.installExtensions(list);
+			
+			List<File> remainingFiles = list.stream().filter(f -> !f.getName().toLowerCase().endsWith(".jar")).collect(Collectors.toList());
+			var supported = ScriptLanguageProvider.getAvailableScriptLanguages().stream().flatMap(l -> Arrays.stream(l.getExtensions())).collect(Collectors.toList());
```

Comment:
This could be troublesome - see javadoc at https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/stream/Collectors.html#toList()
There's no guarantee we'll get a mutable list back, so I think this should be changed.

One way would be to just immediately add the list to a `new HashSet<>()` (there's bound to be a more efficient way to add to an existing collection with streams, but I don't remember it...) - since it's really a set rather than a list that we want anyway.

Reply:
Good catch! I think adding the list to a `HashSet` as you mentioned should fix this. The next commit will fix this.